### ISSUE TITLE
Navigationlink destination disappears immediately after it appears

### ISSUE THREAD

**Issue by ole-kr-stumpf** [pointfreeco/swift-composable-architecture#393]
**Describe the bug**
When a view has exactly 2 NavigationLinks and a `WithViewStore` inside another `WithViewStore`, the destination
for one of the NavigationLink will sometimes disappear right after it appears.

**To Reproduce**
Use the code below. There are two buttons which opens an inner view which has a back button. It is not clear exactly the order to click, but if you click back and forth between the views, eventually one of the inner views will disappear right after it appears.

Zip:
[NavigationLink.zip](https://github.com/pointfreeco/swift-composable-architecture/files/6028134/NavigationLink.zip)

```swift
import SwiftUI
import ComposableArchitecture

@main
struct NavigationLinkApp: App {
    var body: some Scene {
        WindowGroup {
            ContentView(store: Store(initialState: ContentViewState(),
                                   reducer: contentViewReducer,
                                   environment: ContentViewEnv()))
        }
    }
}


struct ContentViewState: Equatable {
    var showInnerView1: Bool = false
    var showInnerView2: Bool = false
}

enum ContentViewAction: Equatable {
    case toggleInnerView1(Bool)
    case toggleInnerView2(Bool)
}

struct ContentViewEnv {}

var contentViewReducer = Reducer<ContentViewState, ContentViewAction, ContentViewEnv> { state, action, env in
    switch action {
    case let .toggleInnerView1(isPresented):
        state.showInnerView1 = isPresented
        return .none
        
    case let .toggleInnerView2(isPresented):
        state.showInnerView2 = isPresented
        return .none
    }
}

struct ContentView: View {
    
    let store: Store<ContentViewState, ContentViewAction>
    
    var body: some View {
        WithViewStore(self.store) { viewStore in
            NavigationView {
                VStack {
                    WithViewStore(self.store) { innerViewStore in
                        
                        NavigationLink(destination: InnerView(),
                                       isActive: viewStore.binding(
                                        get: { $0.showInnerView1 },
                                        send: ContentViewAction.toggleInnerView1)
                        ) {
                            Button(action: {
                                viewStore.send(.toggleInnerView1(true))
                            }) {
                                Text("Go to inner view1")
                            }
                        }
                    }
                    
                    WithViewStore(self.store) { innerViewStore in
                        NavigationLink(destination: InnerView(),
                                       isActive: viewStore.binding(
                                        get: { $0.showInnerView2 },
                                        send: ContentViewAction.toggleInnerView2)
                        ) {
                            Button(action: {
                                viewStore.send(.toggleInnerView2(true))
                            }) {
                                Text("Go to inner view2")
                            }
                        }
                    }
                    
                }
            }
        }
    }
}

struct InnerView: View {
    var body: some View {
        Text("Inner view")
    }
}


```

**Environment**
 - Xcode 12.5 beta 2
 - Swift 5
 - iOS: 14.5 beta 2
 - TCA: Version does not matter, but reproduced on 0.14.0, 0.10.0 and 0.8.0

**Additional context**
Since the bug is introduced in the beta of iOS 14.5, it is likely that this is a SwiftUI bug, but in case this is not fixed in the final release, are there any workarounds, except from merging the two NavigationLinks to one?

**Comment by lukeredpath** [comment]
Does the problem still occur if you remove the outer WithViewStore? I know this is just an example but it serves no purpose here because you don't actually use the outer view store anywhere.

**Comment by ole-kr-stumpf** [comment]
The problem goes away yes. But I guess thats not feasible in a big app, or is nested `WithViewStore`s considered bad practice?

**Comment by jagreenwood** [comment]
I'm having a similar problem but it doesn't appear to have anything to do with nested `WithViewStores`. 

I've posted about it on the swift forums but haven't gotten a response yet: https://forums.swift.org/t/14-5-beta3-navigationlink-unexpected-pop/45279

**Comment by lukeredpath** [comment]
@ole-kr-stumpf there's nothing wrong with nested `WithViewStore` views - your entire app view hierarchy is going to be comprised of nested views with `WithViewStore` views in them. But it is very important to scope your stores correctly so you're not triggering body calculations that aren't needed. 

For example, you don't need to wrap an `IfLetStore` in a `WithViewStore` because it will use its own `WithViewStore` internally to observe changes. You also don't need to use `WithViewStore` if you're not actually using any state, or if you're simply rendering state that was passed in by a parent view which is already observing changes to that state (because the parent will recreate the child view). Also, if you only need access to a view store to send actions, you can use the `.stateless` method to scope your state down to a `Void` state.

**Comment by stephencelis** [comment]
@ole-kr-stumpf Is it OK to close this or would you like me convert it to a discussion? This appears to be more of an issue with how SwiftUI works than a TCA bug, and TCA should indeed aid in eliminating the bug should you `score` your view stores to just the state a particular view cares about.

**Comment by jagreenwood** [comment]
@stephencelis I believe this is the same issue I'm experiencing which I detailed in the forum post above. My primary concern is that vanilla SwiftUI doesn't exhibit this issue. I created a [test project](https://github.com/jagreenwood/tca-nav-pop-bug) to demonstrate the issue.
I would definitely appreciate your expertise in how to avoid the issue with TCA. Currently for any view that _might_ have exactly 2 `NavigationLink`s, I artificially ensure this won't happen by adding the following:

```swift
NavigationLink(destination: EmptyView()) {
    EmptyView()
}
```

As you might agree, this is kinda yuck and doesn't scale well.
I don't have a preference whether this is closed or moved to a discussion, but I do feel some level of support is needed as this is something that will eventually impact anyone who uses TCA with SwiftUI.
